### PR TITLE
Imp/auto local

### PR DESCRIPTION
**Pull request by thebeline**
Attempts to refresh the SD and checks it's state when it's determined it should show root locations.

If the SD is not `ready`, it does one of two things depending on what is occurring:
- If we are just entering the `filesPanel`, it automatically enters `local`.
- If we are iterating backwards out of `local`, it proceeds to the previous panel.

This check happens every time you enter the `filesPanel`, so if you insert an SD card prior to entering the panel, OR insert an SD card while in `local` and go `Back`, it will detect the presence of an SD and display both options.

This fully fleshes the the SD support, as it correctly behaves with the possible `404` response from OctoPrint if [SD Support](https://docs.octoprint.org/en/master/api/printer.html#retrieve-the-current-sd-state) is disabled (which is a thing).

### CODE REVIEW

File: ui/FilesPanel.go [hunk 1]
```diff
@@ -90,33 +90,58 @@ func (this *filesPanel) createRefreshButton() *gtk.Button {
...
+		return false
+	}
+}
+
+func (this *filesPanel) goBack() {
+	if this.isRoot() {
+		this.UI.GoToPreviousPanel()
+	} else if this.locationHistory.IsRoot() {
+		this.locationHistory.GoBack()
+		if this.refreshSD() {
```
Comment (thebeline) [inline ui/FilesPanel.go:133]: It does occur to me that this will cause `refreshSD` to trigger twice (here, as well as in `doLoadFiles`), however there doesn't seem to be any kind of performance penalty from this that I can detect.

File: ui/FilesPanel.go [hunk 2]
```diff
@@ -90,33 +90,58 @@ func (this *filesPanel) createRefreshButton() *gtk.Button {
...
 }
 
 func (this *filesPanel) doLoadFiles() {
 	utils.EmptyTheContainer(&this.listBox.Container)
-
-	if this.displayRootLocations() {
-		this.addRootLocations()
-	} else {
+	if this.isRoot() {
+		if this.refreshSD() {
```
Comment (JeffB42) [inline ui/FilesPanel.go:99]: `if this.refreshSD()` doesn't read well - it's essentially "if command" vs "if query" (or maybe it's just that I've been working with REST APIs and SQL CRUD for way too many years lol...)

The implementation of refreshSD() is essentially, "refresh the SD list and check if it is ready".  I'd like `refreshSD()` to be renamed to, `sdIsReady()`.  This will read better when grokking the code.

File: ui/FilesPanel.go [hunk 3]
```diff
@@ -90,33 +90,58 @@ func (this *filesPanel) createRefreshButton() *gtk.Button {
...
+			this.addRootLocations()
+		} else {
+			this.locationHistory = utils.LocationHistory {
+				Locations: []dataModels.Location{dataModels.Local},
+			}
+		}
+	}
+	if !this.isRoot() {
```
Comment (JeffB42) [inline ui/FilesPanel.go:107]: I like it.  I have a few requests.  When first reading this line, it looks odd.  Line 98 has `if this.isRoot()` an then line 107 has `if !this.isRoot()` and immediately, my thought was, "why is isRoot() being called twice?  Why not just use 'else'?"  Well, reading through the code, I know why, but this wasn't obvious or intuitive.   Please add a comment above line 107, something to the effect of
```
// this.isRoot() is called again because it is driven by locationHistory, which potentially is changed a few lines above.
```

File: ui/FilesPanel.go (hunk 1 above)
Comment (JeffB42) [reply ui/FilesPanel.go:133]: I was going to suggest passing a bool into doLoadFiles(), but it's used as a callback and so it can't be passed in.  It makes an extra REST call, which is expensive.  The next thing which comes to mind is adding a class member, set it before calling doLoadFiles() and then reset it, but I really don't like this as it really pollutes the class for just this one edge case.  Let's leave it in for now.

File: ui/FilesPanel.go [hunk 4]
```diff
@@ -90,33 +90,58 @@ func (this *filesPanel) createRefreshButton() *gtk.Button {
...
+		} else {
+			this.UI.GoToPreviousPanel()
+		}
+	} else {
+		this.locationHistory.GoBack()
+		this.doLoadFiles()
+	}
+}
+
+func (this *filesPanel) isRoot() bool {
```
Comment (JeffB42) [inline ui/FilesPanel.go:144]: I like this function as `displayRootLocations` better than `isRoot`.  The reason is that in `goBack()`, there is:
```
func (this *filesPanel) goBack() {
	if this.isRoot() {
		this.UI.GoToPreviousPanel()
	} else if this.locationHistory.IsRoot() {
		this.locationHistory.GoBack()
```
We have `this.isRoot()` and `this.locationHistory.IsRoot()`.  When you look at the impl of isRoot, the impl is:
```
func (this *filesPanel) isRoot() bool {
	if this.locationHistory.Length() < 1 {
		return true
	} else {
		return false
	}
}
```
...and as a reader, questions come up, like
_"why define isRoot() when it just uses locationHistory.Length()?"_ 
and _"what is locationHistory.Length() vs locationHistory.IsRoot()?"_ 

When named `displayRootLocations()`, it's clearer as to what the intent of the logic is.

Comment (JeffB42) [reply ui/FilesPanel.go:144]: PS. That being said, if you have a better name for `displayRootLocations`, I'm open to it being renamed.

File: ui/FilesPanel.go (hunk 1 above)
Comment (thebeline) [reply ui/FilesPanel.go:133]: I actually did some testing, and it doesn't actually add any perceptible overhead. The menu doesn't hesitate in any noticeable way. Further, doing the call in both directions (and subsequently possibly twice) does account for inserting an SD at any time.

Comment (thebeline) [reply ui/FilesPanel.go:133]: @JeffB42 - So I just ran this side-by-side with another branch with the same feature, there may have been a fraction of a second hesitation, but it didn't really impact the responsiveness or UX.

File: ui/FilesPanel.go (hunk 2 above)
Comment (thebeline) [reply ui/FilesPanel.go:99]: Fair. I called it this because the endpoint in OctoPrint is `refresh`, but you are correct, it is doing more than one thing, so the read reading of the action should reflect that.

File: ui/FilesPanel.go (hunk 4 above)
Comment (thebeline) [reply ui/FilesPanel.go]: I'll go ahead and switch it back.  I am not sure about the name, personally, as it seems like it is an action, rather than a test, but best not to muck with it I guess.

**Comment by thebeline** [comment]
@JeffB42 - I think I sorted your concerns.  Let me know.

File: ui/FilesPanel.go [hunk 5]
```diff
@@ -90,32 +90,62 @@ func (this *filesPanel) createRefreshButton() *gtk.Button {
...
-	})
+	return utils.MustButton(image, this.goBack)
 }
 
 func (this *filesPanel) doLoadFiles() {
 	utils.EmptyTheContainer(&this.listBox.Container)
-
-	if this.displayRootLocations() {
+	atRootLevel := this.displayRootLocations()
+	/**
```
Comment (JeffB42) [inline ui/FilesPanel.go:99]: I'll just clean this up myself, but just a fyi, `/**` is a JSDoc comment (https://jsdoc.app/about-getting-started.html) and typically is only used when generating docs from code.

Comment (thebeline) [reply ui/FilesPanel.go:99]: As an FYI, it may be, but not exclusively:

https://pear.php.net/manual/en/standards.sample.php

That was just the first doc I could pull up, but there are plenty more. PSR-12, etc etc.

I've been programming for 15 years. It's a common enough comment style.

That said, if it goes against the codebase style, I will be sure to adjust when I contribute.

----

A quick look shows that the standard Go convention is:

```
// Single line comment

// Occasionally maybe
// multiple lines.

/*
C-Style block comments.

For when you really have 
to get your comment on.
*/
```

That about right? I'll sort it out next time around. My appologies.

Comment (JeffB42) [reply ui/FilesPanel.go:99]: > As an FYI, it may be, but not exclusively:
> https://pear.php.net/manual/en/standards.sample.php
> That was just the first doc I could pull up, but there are plenty more. PSR-12, etc etc.

TL;DR - I was wrong.  

In several languages (eg JavaScript, Java, and C#, etc...) `/**` is used for auto-generating docs from the source code, and is supposed to be used only for the public-facing/consumable code (things you'd make an API guide from)

GoLang has GoDoc, and I thought it used the same syntax, however:
1) Searching the codebase, I see only a few instances of `/**`, but not everywhere (like you would expect, like in the libraries that are in vendor)
2) The docs for GoLang don't specify it, they say to just use `//`
https://blog.golang.org/godoc

I don't plan on generating and publish docs like this for OctoScreen, but I wanted to keep in line with what the GoLang community does.



> That said, if it goes against the codebase style, I will be sure to adjust when I contribute.

I was wrong.  I thought the doc-generating style was `/**`.  In the future feel free to use this style.  My concern had been that it would add internal comments to docs if they were generated, but that's not the case.



> I've been programming for 15 years. It's a common enough comment style.

Yea, I got the impression you've been programming for a while.  I can tell from your PRs that you're competent, but also impressed that you don't write "clever" code (meaning, some devs like to write complicated, obfuscated code, which no one can read, and a year later when the code needs to be enhanced, no one can change it w/o introducing bugs)

Comment (thebeline) [reply ui/FilesPanel.go:99]: Oh...  I'm only just barely breaking the habit of clever code. lol

I'll just stick to `//` and `/*` for GoLang.  In my experience, `/*` and `/**` have been relatively interchangeable, but it's really neither here nor there.  The best standard is a standard. :-D So I'll stick to the standard.
